File: leetcode/cpp/1370_Increasing_Decreasing_String.cpp
#include <string>
using namespace std;
class Solution {
public:
	string sortString(string s) {
		int freq[26];
		memset(freq, 0, sizeof(freq));

		int cnt = s.length();
		for (char c : s) freq[c - 'a']++;

		string ret;
		while (cnt > 0) {
			for (int i = 0; i < 26; ++i) {
				if (freq[i]) {
					freq[i]--;
					cnt--;
					ret.push_back((char)('a' + i));
				}
			}

			for (int i = 25; i >= 0; --i) {
				if (freq[i]) {
					freq[i]--;
					cnt--;
					ret.push_back((char)('a' + i));
				}
			}
		}

		return ret;
	}
};

File: leetcode/cpp/1370_Increasing_Decreasing_String.py
class Solution:
    def sortString(self, s: str) -> str:
        freq = [0] * 26

        for c in s:
            freq[ord(c) - ord('a')] += 1

        cnt = len(s)
        result = []

        while cnt > 0:
            for i in range(26):
                if cnt == 0:
                    break
                if freq[i] > 0:
                    freq[i] -= 1
                    cnt -= 1
                    result.append(chr(ord('a') + i))

            for i in range(25, -1, -1):
                if cnt == 0:
                    break
                if freq[i] > 0:
                    freq[i] -= 1
                    cnt -= 1
                    result.append(chr(ord('a') + i))

        return "".join(result)
